File: src/game/Combate.js
import { visionGanaA } from "./Vision.js";
import { armaGanaA } from "./Arma.js";

export class Combate {
  static ganaAtacante(atacante, defensor) {
    const elementoAtacante = atacante.getVision();
    const elementoDefensor = defensor.getVision();

    const armaAtacante = atacante.getArma();
    const armaDefensor = defensor.getArma();

    const ventajaElemento = visionGanaA(elementoAtacante, elementoDefensor);
    const desventajaElemento = visionGanaA(elementoDefensor, elementoAtacante);

    const tirada = Math.floor(Math.random() * 10) + 1;

    if (ventajaElemento) return [1,2,3,4,5,6].includes(tirada);
    if (desventajaElemento) return [1,2,3,4].includes(tirada);

    const ventajaArma = armaGanaA(armaAtacante, armaDefensor);
    const desventajaArma = armaGanaA(armaDefensor, armaAtacante);

    if (ventajaArma) return [1,2,3,4,5,6].includes(tirada);
    if (desventajaArma) return [1,2,3,4].includes(tirada);

    return [1,2,3,4,5].includes(tirada);
  }
}

File: java/src/game/Combate.java
package game;

public class Combate {

    public static boolean ganaAtacante(Personaje atacante, Personaje defensor) {

    Vision elementoAtacante = atacante.getVision();
    Vision elementoDefensor = defensor.getVision();

    Arma armaAtacante = atacante.getArma();
    Arma armaDefensor = defensor.getArma();

    boolean ventajaElemento = elementoAtacante.ganaA(elementoDefensor);
    boolean desventajaElemento = elementoDefensor.ganaA(elementoAtacante);

    int tirada = (int)(Math.random() * 10) + 1;

    //Un 60/40 a favor del atacante si tiene ventaja de vision
    //Si el atacante gana, se devuelve true
    if (ventajaElemento) {
        switch (tirada) {
            case 1,2,3,4,5,6:
                return true;
            default:
                return false;
        }
    }
    //Si el defensor gana, se devuelve true
    if (desventajaElemento) {
        switch (tirada) {
            case 1,2,3,4:
                return true; 
            default:
                return false;
        }
    }

    boolean ventajaArma = armaAtacante.ganaA(armaDefensor);
    boolean desventajaArma = armaDefensor.ganaA(armaAtacante);

    //Un 60/40 a favor del atacante si tiene ventaja de arma
    //Si el atacante gana, se devuelve true
    if (ventajaArma) {
        switch (tirada) {
            case 1,2,3,4,5,6:
                return true;
            default:
                return false;
        }
    }

    //Si el defensor gana, se devuelve true
    if (desventajaArma) {
        switch (tirada) {
            case 1,2, 3, 4:
                return true;
            default:
                return false;
        }
    }

    //50/50 en caso de empate en armas y visiones
    switch (tirada) {
        case 1,2,3,4,5:
            return true;
        default:
            return false;
    }
}

}
